File: backend/app/api/commands.py
from __future__ import annotations

import sqlite3

from fastapi import APIRouter, Depends

from app.core.database import get_db
from app.domain.enums import StageId, Status
from app.domain.errors import NotFoundError, ValidationError
from app.domain.schemas import (
    CommandCreateRequest,
    CommandReorderRequest,
    CommandResponse,
    CommandUpdateRequest,
)
from app.repositories.command_repository import CommandRepository
from app.services.command_service import CommandService

router = APIRouter()


def _service(conn: sqlite3.Connection) -> CommandService:
    return CommandService(CommandRepository(conn))


@router.get("/api/commands", response_model=list[CommandResponse])
def list_commands(
    stage_id: str | None = None,
    status: str | None = None,
    conn: sqlite3.Connection = Depends(get_db),
) -> list[CommandResponse]:
    parsed_stage_id = None
    if stage_id is not None:
        parsed_stage_id = StageId.from_str(stage_id)
        if parsed_stage_id is None:
            raise ValidationError("Invalid stage_id")

    parsed_status = None
    if status is not None:
        parsed_status = Status.from_str(status)
        if parsed_status is None:
            raise ValidationError("Invalid status")

    service = _service(conn)
    return [
        CommandResponse.from_model(c)
        for c in service.list(parsed_stage_id, parsed_status)
    ]


@router.post("/api/commands", response_model=CommandResponse, status_code=201)
def create_command(
    payload: CommandCreateRequest,
    conn: sqlite3.Connection = Depends(get_db),
) -> CommandResponse:
    parsed_stage_id = StageId.from_str(payload.stage_id)
    if parsed_stage_id is None:
        raise ValidationError("Invalid stage_id")

    parsed_status = Status.NOT_STARTED
    if payload.status is not None:
        maybe_status = Status.from_str(payload.status)
        if maybe_status is None:
            raise ValidationError("Invalid status")
        parsed_status = maybe_status

    service = _service(conn)
    cmd = service.create(
        title=payload.title,
        stage_id=parsed_stage_id,
        status=parsed_status,
    )
    return CommandResponse.from_model(cmd)


@router.get("/api/commands/{command_id}", response_model=CommandResponse)
def get_command(
    command_id: int,
    conn: sqlite3.Connection = Depends(get_db),
) -> CommandResponse:
    service = _service(conn)
    cmd = service.get(command_id)
    if cmd is None:
        raise NotFoundError("Command not found")
    return CommandResponse.from_model(cmd)


@router.patch("/api/commands/{command_id}", response_model=CommandResponse)
def update_command(
    command_id: int,
    payload: CommandUpdateRequest,
    conn: sqlite3.Connection = Depends(get_db),
) -> CommandResponse:
    parsed_stage_id = None
    if payload.stage_id is not None:
        parsed_stage_id = StageId.from_str(payload.stage_id)
        if parsed_stage_id is None:
            raise ValidationError("Invalid stage_id")

    parsed_status = None
    if payload.status is not None:
        parsed_status = Status.from_str(payload.status)
        if parsed_status is None:
            raise ValidationError("Invalid status")

    service = _service(conn)
    cmd = service.update(
        command_id,
        title=payload.title,
        stage_id=parsed_stage_id,
        status=parsed_status,
    )
    if cmd is None:
        raise NotFoundError("Command not found")
    return CommandResponse.from_model(cmd)


@router.delete("/api/commands/{command_id}")
def delete_command(
    command_id: int,
    conn: sqlite3.Connection = Depends(get_db),
) -> dict[str, bool]:
    service = _service(conn)
    ok = service.delete(command_id)
    if not ok:
        raise NotFoundError("Command not found")
    return {"ok": True}


@router.post("/api/commands/reorder")
def reorder_commands(
    payload: CommandReorderRequest,
    conn: sqlite3.Connection = Depends(get_db),
) -> dict[str, bool]:
    """Persist a new ordering for one or more categories.

    The payload provides ordered command id lists keyed by category name.
    """

    by_stage_id: dict[StageId, list[int]] = {}
    for raw_stage_id, ids in payload.by_stage_id.items():
        parsed = StageId.from_str(raw_stage_id)
        if parsed is None:
            raise ValidationError("Invalid stage_id")
        by_stage_id[parsed] = ids

    service = _service(conn)
    service.reorder(by_stage_id)
    return {"ok": True}

File: frontend/src/api/commands.ts
import { apiFetch } from "./http";

export type StageId = "DESIGN" | "BUILD" | "REVIEW" | "COMPLETE";
export type Status = "Not Started" | "In Progress" | "Blocked" | "Complete";

export type Command = {
  id: number;
  title: string;
  stage_id: StageId;
  status: Status;
  created_at: string;
};

export type CreateCommandInput = {
  title: string;
  stage_id: StageId;
  status?: Status;
};

export type UpdateCommandInput = {
  title?: string;
  stage_id?: StageId;
  status?: Status;
};

export async function listCommands(params?: {
  stage_id?: StageId;
  status?: Status;
}): Promise<Command[]> {
  const qs = new URLSearchParams();
  if (params?.stage_id) qs.set("stage_id", params.stage_id);
  if (params?.status) qs.set("status", params.status);

  const path = qs.toString() ? `/api/commands?${qs.toString()}` : "/api/commands";
  return await apiFetch<Command[]>(path);
}

export async function createCommand(input: CreateCommandInput): Promise<Command> {
  return await apiFetch<Command>("/api/commands", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export async function updateCommand(
  id: number,
  input: UpdateCommandInput
): Promise<Command> {
  return await apiFetch<Command>(`/api/commands/${id}`, {
    method: "PATCH",
    body: JSON.stringify(input),
  });
}

export async function deleteCommand(id: number): Promise<{ ok: true }> {
  return await apiFetch<{ ok: true }>(`/api/commands/${id}`, {
    method: "DELETE",
  });
}

export async function reorderCommands(input: {
  by_stage_id: Record<StageId, number[]>;
}): Promise<{ ok: true }> {
  return await apiFetch<{ ok: true }>("/api/commands/reorder", {
    method: "POST",
    body: JSON.stringify(input),
  });
}
